### PR TITLE
Change logic of stdin input to be last and only if required

### PR DESCRIPTION
**Pull request by Kubuxu**
This PR changes how stdin is used while parsing arguments. Current behaviour is that stdin is used for non-required and variadic arguments which causes confusion.

After the change the stdin is used only in case of missing required argument or if explicitly mentioned by the user with `-`.

Question: Should we print on stderr when ipfs command is listening for stdin to clear up possible confusing hang even more?

Aims to resolve https://github.com/ipfs/go-ipfs/issues/2748

License: MIT
Signed-off-by: Jakub Sztandera <kubuxu@protonmail.ch>

### CODE REVIEW

**Comment by whyrusleeping** [comment]
@Kubuxu could you provide a couple examples of how this changes command line usage for the end user? Does it have any effect on things like `echo "test" | ipfs add` ?

**Comment by Kubuxu** [comment]
No it doesn't for the most part.

It changes behaviour if parameter is optional user as has to explicitly specify that he wants to use stdin using `-` argument. So it changes semantic of for example `ipfs init` where the file arugment is optional. To initialize repo from stdin one has to do `echo '{... config ... } | ipfs init -`.

As argument for add is required it will wait for input. I think we should add a message on stderr about it as currently `ipfs add` will display help, with this patch it will silently wait for user input.

Something along the lines of:
`Reading standard input, <Ctrl-D> to finish. $CMD_PATH -h for more information.`
if input is a terminal. It would greatly reduce confusion when command doesn't return.

@RichardLitt can you check wording of that?

File: test/bin/time-out [hunk 1]
```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
+#
+# The Bash shell script executes a command with a time-out.
+# Upon time-out expiration SIGTERM (15) is sent to the process. If the signal
+# is blocked, then the subsequent SIGKILL (9) terminates it.
+#
+# Based on the Bash documentation example.
+
+# Hello Chet,
```
Comment (RichardLitt) [inline test/bin/time-out]: What's this?

File: test/bin/time-out [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+# please find attached a "little easier"  :-)  to comprehend
+# time-out example.  If you find it suitable, feel free to include
+# anywhere: the very same logic as in the original examples/scripts, a
+# little more transparent implementation to my taste.
+#
+# Dmitry V Golovashkin <Dmitry.Golovashkin@sas.com>
```
Comment (RichardLitt) [inline test/bin/time-out]: Who is this?

**Comment by RichardLitt** [comment]
Looks fine to me from a wording perspective. Not sure what that shell script is.

File: test/bin/time-out (hunk 2 above)
Comment (Kubuxu) [inline test/bin/time-out]: It is from a website that was linked on SO, it was released as a free (public domain).
The script in whole is for running commands with timeout.

Comment (RichardLitt) [inline test/bin/time-out]: Still, those comments aren't necessary, are they? I would remove them.

**Comment by whyrusleeping** [comment]
tried this out, works good on my end. makes docker a bit happier.

Thanks @Kubuxu !

File: test/bin/time-out [hunk 3]
```diff
@@ -0,0 +1,83 @@
+#!/bin/bash
+#
+# The Bash shell script executes a command with a time-out.
+# Upon time-out expiration SIGTERM (15) is sent to the process. If the signal
+# is blocked, then the subsequent SIGKILL (9) terminates it.
+#
+# Based on the Bash documentation example.
```
Comment (jbenet) [inline test/bin/time-out]: @Kubuxu did you write this? if not, what is the license?

**Comment by jbenet** [comment]
IIRC, this PR introduced a bunch of complexity and breaking changes. Have these been resolved? if so, where? (this may be ahead in my CR)

**Comment by jbenet** [comment]
Found it https://github.com/ipfs/go-ipfs/pull/2902

File: test/bin/time-out (hunk 3 above)
Comment (Kubuxu) [inline test/bin/time-out]: It is Public Domain, I had previously credited the original author but removed it with other comments as for @RichardLitt 's request. I shouldn't have removed the author line, it was a signature for a short letter there.

**Comment by jbenet** [comment]
More here https://github.com/ipfs/go-ipfs/pull/2952/files

**Comment by jbenet** [comment]
would be good to isolate all these commits and see a diff between `v0.4.2..v0.4.3-rc3` with just all of the commits of these PRs. to ensure things got reverted appropriately

**Comment by lgierth** [comment]
> would be good to isolate all these commits and see a diff between v0.4.2..v0.4.3-rc3 with just all of the commits of these PRs. to ensure things got reverted appropriately

I tried to somewhat group the 0.4.3-rc1 changelog entries logically, but the ones for stdin didn't turn out that well :) some entries are pretty good though and we should continue to aim for properly editorialized changelogs (i.e. logically grouped and structured)

**Comment by lgierth** [comment]
> I tried to somewhat group the 0.4.3-rc1 changelog entries logically, but the ones for stdin didn't turn out that well :) some entries are pretty good though and we should continue to aim for properly editorialized changelogs (i.e. logically grouped and structured)

tl;dr merges don't map to changelog entries

File: test/bin/time-out (hunk 3 above)
Comment (jbenet) [inline test/bin/time-out]: Author line needs to be back before release.

Comment (Kubuxu) [inline test/bin/time-out]: This should address the license issue: https://github.com/ipfs/go-ipfs/pull/3152

File: test/bin/time-out [hunk 4]
```diff
@@ -0,0 +1,83 @@
+#!/bin/bash
```
Comment (chriscool) [inline test/bin/time-out]: I am a bit concerned that this script is written in bash.

File: test/bin/time-out [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    -i interval
+        Interval between checks if the process is still alive.
+        Positive integer, default value: $DEFAULT_INTERVAL seconds.
+
+    -d delay
+        Delay between posting the SIGTERM signal and destroying the
+        process by SIGKILL. Default value: $DEFAULT_DELAY seconds.
+
+As of today, Bash does not support floating point arithmetic (sleep does),
+therefore all delay/time values must be integers.
```
Comment (chriscool) [inline test/bin/time-out]: We have go-sleep that accepts floating point delays.

File: test/bin/time-out (hunk 4 above)
Comment (Kubuxu) [inline test/bin/time-out]: @chriscool we removed this script in https://github.com/ipfs/go-ipfs/pull/3152

Comment (chriscool) [inline test/bin/time-out]: Yeah, I saw it afterwards. Thanks for having written something more portable!
